### PR TITLE
use precomputed border values

### LINKED ISSUE #15142: Border radius is scaled incorrectly
## Bevy version

main

## What you did

Modified button example:

```rust
use bevy::{color::palettes::basic::*, prelude::*, winit::WinitSettings};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        // Only run the app when there is user input. This will significantly reduce CPU/GPU use.
        .insert_resource(UiScale(2.))
        .insert_resource(WinitSettings::desktop_app())
        .add_systems(Startup, setup)
        .add_systems(Update, button_system)
        .run();
}

const NORMAL_BUTTON: Color = Color::srgb(0.15, 0.15, 0.15);
const HOVERED_BUTTON: Color = Color::srgb(0.25, 0.25, 0.25);
const PRESSED_BUTTON: Color = Color::srgb(0.35, 0.75, 0.35);

fn button_system(
    mut interaction_query: Query<
        (
            &Interaction,
            &mut BackgroundColor,
            &mut BorderColor,
            &Children,
        ),
        (Changed<Interaction>, With<Button>),
    >,
    mut text_query: Query<&mut Text>,
) {
    for (interaction, mut color, mut border_color, children) in &mut interaction_query {
        let mut text = text_query.get_mut(children[0]).unwrap();
        match *interaction {
            Interaction::Pressed => {
                text.sections[0].value = "Press".to_string();
                *color = PRESSED_BUTTON.into();
                border_color.0 = RED.into();
            }
            Interaction::Hovered => {
                text.sections[0].value = "Hover".to_string();
                *color = HOVERED_BUTTON.into();
                border_color.0 = Color::WHITE;
            }
            Interaction::None => {
                text.sections[0].value = "Button".to_string();
                *color = NORMAL_BUTTON.into();
                border_color.0 = Color::BLACK;
            }
        }
    }
}

fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
    // ui camera
    commands.spawn(Camera2dBundle::default());
    commands
        .spawn(NodeBundle {
            style: Style {
                width: Val::Percent(100.0),
                height: Val::Percent(100.0),
                align_items: AlignItems::Center,
                justify_content: JustifyContent::Center,
                ..default()
            },
            ..default()
        })
        .with_children(|parent| {
            parent
                .spawn(ButtonBundle {
                    style: Style {
                        width: Val::Px(250.0),
                        height: Val::Px(100.0),
                        border: UiRect::all(Val::Px(25.0)),
                        // horizontally center child text
                        justify_content: JustifyContent::Center,
                        // vertically center child text
                        align_items: AlignItems::Center,
                        ..default()
                    },
                    border_color: BorderColor(Color::BLACK),
                    border_radius: BorderRadius::all(Val::Px(25.)),
                    background_color: NORMAL_BUTTON.into(),
                    ..default()
                })
                .with_child(TextBundle::from_section(
                    "Button",
                    TextStyle {
                        font: asset_server.load("fonts/FiraSans-Bold.ttf"),
                        font_size: 40.0,
                        color: Color::srgb(0.9, 0.9, 0.9),
                    },
                ));
            parent
                .spawn(NodeBundle {
                    style: Style {
                        width: Val::Px(150.0),
                        padding: UiRect::vertical(Val::Px(25.)),
                        height: Val::Px(100.0),
                        align_items: AlignItems::Stretch,
                        justify_content: JustifyContent::Stretch,
                        ..Default::default()
                    },
                    ..default()
                })
                .with_child(NodeBundle {
                    style: Style {
                        flex_basis: Val::Percent(100.),
                        ..Default::default()
                    },
                    background_color: RED.into(),
                    ..Default::default()
                });
        });
}
```

## What went wrong

With `UiScale(1.)`:

br1

With `UiScale(2.)`:
br2

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Fixes #15142

## Solution

* Moved all the UI border geometry calculations that were scattered through the UI extraction functions into `ui_layout_system`.
* Added a `border: BorderRect` field to `Node` to store the border size computed by `ui_layout_system`.
* Use the border values returned from Taffy rather than calculate them ourselves during extraction.
* Removed the `logical_rect` and `physical_rect` methods from `Node` the descriptions and namings are deceptive, it's better to create the rects manually instead.
* Added a method `outline_radius` to `Node` that calculates the border radius of outlines.
* For border values `ExtractedUiNode` takes `BorderRect` and `ResolvedBorderRadius` now instead of raw `[f32; 4]` values and converts them in `prepare_uinodes`.
* Removed some unnecessary scaling and clamping of border values (#15142).
* Added a `BorderRect::ZERO` constant.
* Added an `outlined_node_size` method to `Node`.

## Testing

Added some non-uniform borders to the border example. Everything seems to be in order:

nub

## Migration Guide

The `logical_rect` and `physical_rect` methods have been removed from `Node`. Use `Rect::from_center_size` with the translation and node size instead.

The types of the fields border and border_radius of `ExtractedUiNode` have been changed to `BorderRect` and `ResolvedBorderRadius` respectively.

### CODE REVIEW

File: crates/bevy_ui/src/render/ui.wgsl [hunk 1]
```diff
@@ -156,6 +164,12 @@ fn draw(in: VertexOutput, texture_color: vec4<f32>) -> vec4<f32> {
     // a border.
     let t = select(1.0 - step(0.0, border_distance), antialias(border_distance), external_distance < internal_distance);
 
+
+    if is_inside_corner(in.point, in.size, in.radius) {
+        return vec4(0., 1., 0., t);
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/ui.wgsl:169]: Reminder to self: ensure this code is reverted.

File: crates/bevy_ui/src/render/mod.rs [hunk 2]
```diff
@@ -403,67 +310,28 @@ pub fn extract_uinode_images(
                 flip_x: image.flip_x,
                 flip_y: image.flip_y,
                 camera_entity,
-                border,
-                border_radius,
+                border: BorderRect::ZERO,
+                border_radius: ResolvedBorderRadius::ZERO,
                 node_type: NodeType::Rect,
```
Comment (UkoeHB) [inline crates/bevy_ui/src/render/mod.rs:315]: Intentional? Add a comment explaining

File: crates/bevy_ui/src/render/mod.rs [hunk 3]
```diff
@@ -487,51 +353,14 @@ pub fn extract_uinode_borders(
 
         // Skip invisible borders
         if !view_visibility.get()
-            || style.display == Display::None
             || maybe_border_color.is_some_and(|border_color| border_color.0.is_fully_transparent())
```
Comment (UkoeHB) [inline crates/bevy_ui/src/render/mod.rs:356]: Looks unrelated

File: crates/bevy_ui/src/render/mod.rs [hunk 4]
```diff
@@ -487,51 +353,14 @@ pub fn extract_uinode_borders(
...
-            uinode.border_radius.top_left,
-            uinode.border_radius.top_right,
-            uinode.border_radius.bottom_right,
-            uinode.border_radius.bottom_left,
-        ]
-        .map(|r| r * ui_scale.0);
-
-        let border_radius = clamp_radius(border_radius, uinode.size(), border.into());
-
+        //let border_radius = clamp_radius(border_radius, uinode.size(), border.into());
```
Comment (UkoeHB) [inline crates/bevy_ui/src/render/mod.rs:361]: Leftover?

File: crates/bevy_ui/src/render/mod.rs [hunk 5]
```diff
@@ -559,13 +388,6 @@ pub fn extract_uinode_borders(
...
         if let Some(outline) = maybe_outline {
             let outer_distance = uinode.outline_offset() + uinode.outline_width();
-            let outline_radius = border_radius.map(|radius| {
-                if radius > 0. {
-                    radius + outer_distance
-                } else {
-                    0.
-                }
-            });
             let outline_size = uinode.size() + 2. * outer_distance;
```
Comment (UkoeHB) [inline crates/bevy_ui/src/render/mod.rs:391]: `outline_size` could be moved to a `Node` method too while you're at it.

File: crates/bevy_ui/src/render/ui.wgsl [hunk 6]
```diff
@@ -93,6 +93,14 @@ fn sd_rounded_box(point: vec2<f32>, size: vec2<f32>, corner_radii: vec4<f32>) ->
     return l + m - radius;
 }
 
+fn is_inside_corner(point: vec2<f32>, size: vec2<f32>, corner_radii: vec4<f32>) -> bool {
```
Comment (UkoeHB) [inline crates/bevy_ui/src/render/ui.wgsl:96]: If keeping this, add comment that it's currently available for debugging and not used for anything.

**Review by UkoeHB — CHANGES_REQUESTED** [review]
Looking good. Can you update the example to include varied combinations of border width and border radius? Looks like it's all uniform right now.

File: crates/bevy_ui/src/render/ui.wgsl [hunk 7]
```diff
@@ -171,10 +185,12 @@ fn draw_background(in: VertexOutput, texture_color: vec4<f32>) -> vec4<f32> {
 
 @fragment
 fn fragment(in: VertexOutput) -> @location(0) vec4<f32> {
-    let texture_color = textureSample(sprite_texture, sprite_sampler, in.uv);
+
+
+    var texture_color = textureSample(sprite_texture, sprite_sampler, in.uv);
```
[the remark this answers is not in the document]
Comment (ickshonpe) [reply crates/bevy_ui/src/render/ui.wgsl:190]: I have an updated borders example I was working on but I'd rather put it in as a separate PR.

File: crates/bevy_ui/src/render/mod.rs (hunk 5 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:391]: Yeah, called the method `outlined_node_size` as `outline_size` could be confusing.

File: crates/bevy_ui/src/render/ui.wgsl (hunk 6 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/ui.wgsl:96]: Yep not going to keep it.

File: crates/bevy_ui/src/render/mod.rs (hunk 2 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:315]: Ah yeah I mixed this up, it's incorrect.

File: crates/bevy_ui/src/render/mod.rs (hunk 3 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:357]: Removing this check means we don't have to query for `Style`, With  `Display::None` the thickness is zero so it still gets skipped.

**Comment by ickshonpe** [comment]
Okay this should be ready now.

**Review by UkoeHB — APPROVED** [review]

**Review by kristoff3r — APPROVED** [review]
